spring-batch-core/src/main/java/org/springframework/batch/core/partition/support/SimpleStepExecutionSplitter.java

```diff
@@ -178,14 +178,17 @@ public String getStepName() {
 		Map<String, ExecutionContext> contexts = getContexts(stepExecution, gridSize);
 		Set<StepExecution> set = new HashSet<>(contexts.size());
 
+		JobInstance jobInstance = stepExecution.getJobExecution().getJobInstance();
+		Collection<StepExecution> allPriorStepExecutions = jobRepository.getStepExecutions(jobInstance);
+
 		for (Entry<String, ExecutionContext> context : contexts.entrySet()) {
 
 			// Make the step execution name unique and repeatable
 			String stepName = this.stepName + STEP_NAME_SEPARATOR + context.getKey();
 
 			StepExecution currentStepExecution = jobExecution.createStepExecution(stepName);
 
-			boolean startable = isStartable(currentStepExecution, context.getValue());
+			boolean startable = isStartable(currentStepExecution, context.getValue(), allPriorStepExecutions);
```

Comment:
`isStartable` requires the current step execution as well as the latest one with the same name. So there is no need to pass the entire collection of step executions here. You can get the last one with the protected utility method `getLastStepExecution` and pass it to `isStartable`.